examples/practical-coalton/15_pathnames.lisp

```diff
@@ -0,0 +1,163 @@
+(in-package :cl-user)
+(defpackage :practical-coalton.pathnames
+  (:use
+   #:coalton
+   #:coalton-prelude)
+  (:local-nicknames
+    (#:pth #:com.gigamonkeys.pathnames)
+    (#:str #:coalton-library/string))
+  (:export
+   :list-directory
+   :file-exists-p
+   :directory-pathname-p
+   :file-pathname-p
+   :pathname-as-directory
+   :pathname-as-file
+   :walk-directory
+   :directory-p
+   :file-p
+   :IncludeDirs
+   :ExcludeDirs))
+(in-package :practical-coalton.pathnames)
+
+(named-readtables:in-readtable coalton:coalton)
+
+;; Port of Chapter 15 from Practical Common Lisp, by Peter Seibel,
+;; to Coalton.
+;;
+;; In the previous chapter, we took the approach of rewriting the
+;; Common Lisp functions in Coalton. In this chapter, we instead
+;; write a thin wrapper around the Common Lisp functions.
+;;
+;; First, we start by defining a few Coalton types that provide
+;; a type-safe interface to the Common Lisp pathname functions.
+;;
+;; Second, we define a corresponding Coalton function for each
+;; function exported by the Common Lisp pathname package.
+
+(cl:defmacro wrap-lisp-call (type function value)
+  "Wrap a call to a Lisp function that returns a value of type TYPE
+in a Coalton function that returns a value of type (Optional TYPE).
+A nil return value is converted to None."
+  `(lisp (Optional ,type) (,value)
+     (cl:let ((result (,function ,value)))
+       (cl:if result
+         (coalton (Some (lisp ,type () result)))
+         (coalton None)))))
+
+(coalton-toplevel
+
+  (repr :native cl:pathname)
+  (define-type Pathname)
+
+  (declare pathname->string (Pathname -> String))
+  (define (pathname->string pathname)
+    (lisp String (pathname)
+      (cl:write-to-string pathname)))
+  
+  (repr :native cl:pathname)
+  (define-type WildPathname)
+
+  ;; TODO: Is this safe? I want to be able to pass any of these
```

Comment:
Is this kosher? It saves a lot of typing, but I'm not sure if `(repr :native cl:t)` is dangerous to use for some reason.